theano/gpuarray/opt.py

```diff
@@ -215,6 +254,263 @@ def apply(self, fgraph):
+        for opt in self.local_optimizers_all:
+            yield opt
+        # if repeat is not a problem we can drop the set
+        s = set()
+        for lopt in itervalues(self.local_optimizers_map):
+            for opt in lopt:
+                if opt not in s:
+                    yield opt
+                    s.add(opt)
+
+    def apply(self, fgraph):
+        mapping = {}
+        start_nb_nodes = len(fgraph.apply_nodes)
+        max_nb_nodes = len(fgraph.apply_nodes)
+        io_toposort_timing = []
+        nb_nodes = []
+        time_opts = {}
+        node_created = {}
+        process_count = {}
+        # Building a new graph
+        # Iterating through inputs of graph
+        for i in fgraph.inputs:
+            if isinstance(i.type, tensor.TensorType):
+                mapping[i] = as_gpuarray_variable(i, None)  # TODO context
+            else:
+                mapping[i] = i
+        for i in fgraph.variables:
+            if isinstance(i, theano.Constant):
+                mapping[i] = i
+        for node in fgraph.toposort():
+            for lopt in (self.local_optimizers_all + 
+                         self.local_optimizers_map.get(type(node.op), []) +
+                         self.local_optimizers_map.get(node.op, [])):
+                    process_count.setdefault(lopt, 0)
+                    time_opts.setdefault(lopt, 0)
+                    node_created.setdefault(lopt, 0)
+
+        t_topo = time.time()
+        topo = fgraph.toposort()
+        time_topo = time.time() - t_topo
+
+        for node in fgraph.toposort():
+
+            t0 = time.time()
+            if isinstance(node.op, HostFromGpu):
+                mapping[node.outputs[0]] = node.inputs[0]
+                continue
+
+            # Move only if any of the inputs are on the GPU.
+            move_to_GPU = True
+
+            '''
+            if any([isinstance(i, GpuArrayVariable) or
+                    isinstance(i, GpuArraySharedVariable)
+                    for i in [mapping[v] for v in node.inputs] +
+                              node.outputs]):
+
+                move_to_GPU = True
+            '''
+
+            out_clients = [o.clients for o in node.outputs]
+
+            context_name = None
+            for i in [mapping[i] for i in node.inputs]:
+                if isinstance(i.type, GpuArrayType):
+                    context_name = i.type.context_name
+                    move_to_GPU = True
+                    break
+
+            new_ops = None
+            outputs = []
+            # Apply the lifter
+            for lopt in (self.local_optimizers_all +
+                         self.local_optimizers_map.get(type(node.op), []) +
+                         self.local_optimizers_map.get(node.op, [])):
+                
+                process_count[lopt] += 1
+                if move_to_GPU:
+                    node_created[lopt] += len(theano.gof.graph.ops([mapping[i] for i in node.inputs], node.outputs))
```

Comment:
This line is probably the cause of the slow down. I'll try to find the right fix.